base/path.jl

```diff
@@ -390,7 +390,19 @@ normpath(a::AbstractString, b::AbstractString...) = normpath(joinpath(a,b...))
 Convert a path to an absolute path by adding the current directory if necessary.
 Also normalizes the path as in [`normpath`](@ref).
 """
-abspath(a::String) = normpath(isabspath(a) ? a : joinpath(pwd(),a))
+function abspath(a::String)::String
+    if !Sys.iswindows()
+        return normpath(isabspath(a) ? a : joinpath(pwd(),a))
+    else
+        pwd_drive, _ = splitdrive(pwd())
+        a_drive, a_nodrive = splitdrive(a)
+        if a_drive != "" && lowercase(pwd_drive) != lowercase(a_drive) && !isabspath(a)
+            return a_drive * normpath(joinpath(path_separator, a_nodrive))
+        else
+            return normpath(isabspath(a) ? a : joinpath(pwd(),a))
+        end
+    end
```

Comment:
I think there's some opportunity to minimize code duplication, as follows. This also handles looking at ENV when necessary, matching my observations of how `realpath` would handle these paths.
```suggestion
    if !isabspath(a)
        cwd = pwd()
        a_drive, a_nodrive = splitdrive(a)
        if a_drive != "" && lowercase(splitdrive(cwd)[1]) != lowercase(a_drive)
            cwd = get(ENV, "=" * a_drive, nothing) # check DOS directory
            if cwd === nothing || !endswith(cwd, '\\') || !isabspath(cwd)
                cwd = a_drive * path_separator
            end
            a = joinpath(cwd, a_nodrive)
        else
            a = joinpath(cwd, a)
        end
    end
    return normpath(a)
```

Reply:
Why do we have to bother with DOS directories?

Reply:
It would also be good to avoid   accessing environment variables, unless necessary.

Reply:
@vtjnash  I just rereviewed your suggested changes.  I like the proposed updates, as they consolidate the code and is more readable. 

My only concern, do we really need to bother with the DOS checks (at a quick search I couldn't find a reference on them)? If they are not necessary, I think it would be good to remove them as it would avoid any issues checking environment variables with, e.g., race conditions etc. 
```
            cwd = get(ENV, "=" * a_drive, nothing) # check DOS directory
	            if cwd === nothing || !endswith(cwd, '\\') || !isabspath(cwd)
	                cwd = a_drive * path_separator
	            end
```

Reply:
We could assume `cwd = nothing` there (nothing in ENV). It won't match the Win32 system behavior, but it wouldn't necessarily be wrong either, since we don't claim that this function does.